// PR c++/13269
// { dg-options "-fshow-column" }
// { dg-additional-options "-Wno-return-type" }

class Foo { int foo() return 0; } };

// { dg-error "30:expected identifier before numeric constant" "identifier" { target *-*-* } 5 }

// { dg-error "23:named return values are no longer supported" "named return" { target *-*-* } 5 }

// the column number info of this error output is still wrong because the error
// message has been generated by cp_parser_error() which does not
// necessarily allow accurate column number display. At some point, we will
// need make cp_parser_error() report more accurate column numbers.
// { dg-error "30:expected '\{' at end of input" "brace" { target *-*-* } 5 }

// { dg-error "34:expected ';' after class definition" "semicolon" {target *-*-* } 5 }

// { dg-error "35:expected declaration before '\}' token" "declaration" {target *-*-* } 5 }
